openedx/openedx-platform
# feat: User agreements API for generic agreement records

## Description

This change adds a new kind of generic user agreement that allows plugins or even the core platform to record a user's acknowledgement of an agreement.

## Supporting information

Product Review: https://github.com/openedx/platform-roadmap/issues/465

## Testing instructions

- This creates a new REST APIs for managing agreements:
- The API at /api/agreements/v1/agreement/ will list user agreements.  /api/agreements/v1/agreement/<type>/ will return details about an agreement of a particular type and  /api/agreements/v1/agreement/<type>/text will return the full licence / agreement text
-  The API at /api/agreements/v1/agreement_record/<agreement_type> will return the current user's agreement acceptance status. POSTing to it will cause that agreement type to be accepted. 


To test with the [corresponding frontend change](https://github.com/openedx/frontend-app-authoring/pull/2845) you need a few things:
- Create a user agreement using the django admin
- Use the MFE config API to configure gating based on acceptance as follows:
```json
{
  "AGREEMENT_GATING": {
    "uploads.video": ["abc_xyz"]
  }
}
```
This is essentially saying the to access the "uploads.video" feature you need to accept an agreement with the type "abc_xyz" so fill in the id from the agreement created before. 

Now the UI should show an alert to accept the agreement, and the video upload UI will be blurred out and blocked till the agreement is accepted. The other gating options are "upload.files" and "uploads". The idea being that if a user accepts an agreement that unblocks uploads it will automatically unblock files and videos but there is the option of being more granular.

## Deadline

"None" 


Private-ref: https://tasks.opencraft.com/browse/BB-10395
